include/boost/random/faure.hpp

```diff
@@ -0,0 +1,342 @@
+ * http://www.boost.org/LICENSE_1_0.txt)
+ */
+
+#ifndef BOOST_RANDOM_FAURE_HPP
+#define BOOST_RANDOM_FAURE_HPP
+
+#include <boost/random/detail/qrng_base.hpp>
+
+#include <vector>
+#include <algorithm>
+
+#include <cmath>
+
+#include <boost/assert.hpp>
+#include <boost/multiprecision/integer.hpp> // powm
+
+//!\file
+//!Describes the quasi-random number generator class template faure.
+
+namespace boost {
+namespace random {
+
+/** @cond */
+namespace detail {
+namespace fr {
+
+// There is no particular reason why 187 first primes were chosen
+// to be put into this table. The only reason was, perhaps, that
+// the number of dimensions for Faure generator would be around
+// the same number as the number of dimensions supported by the
+// Sobol qrng.
+struct prime_table
+{
+  typedef unsigned short value_type;
+
+  BOOST_STATIC_CONSTANT(int, number_of_primes = 187);
+
+  // A function that returns lower bound prime for a given n
+  static value_type lower_bound(std::size_t n)
+  {
+    static const value_type prim_a[number_of_primes] = {
+      2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53,
+      59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113,
+      127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181,
+      191, 193, 197, 199, 211, 223, 227, 229, 233, 239, 241, 251,
+      257, 263, 269, 271, 277, 281, 283, 293, 307, 311, 313, 317,
+      331, 337, 347, 349, 353, 359, 367, 373, 379, 383, 389, 397,
+      401, 409, 419, 421, 431, 433, 439, 443, 449, 457, 461, 463,
+      467, 479, 487, 491, 499, 503, 509, 521, 523, 541, 547, 557,
+      563, 569, 571, 577, 587, 593, 599, 601, 607, 613, 617, 619,
+      631, 641, 643, 647, 653, 659, 661, 673, 677, 683, 691, 701,
+      709, 719, 727, 733, 739, 743, 751, 757, 761, 769, 773, 787,
+      797, 809, 811, 821, 823, 827, 829, 839, 853, 857, 859, 863,
+      877, 881, 883, 887, 907, 911, 919, 929, 937, 941, 947, 953,
+      967, 971, 977, 983, 991, 997, 1009, 1013, 1019, 1021, 1031,
+      1033, 1039, 1049, 1051, 1061, 1063, 1069, 1087, 1091, 1093,
+      1097, 1103, 1109, 1117 };
+
+    detail::dimension_assert("Faure", n, prim_a[number_of_primes - 1]);
+
+    return *std::lower_bound(prim_a, prim_a + number_of_primes, n);
+  }
+};
+
+// Returns the integer part of the logarithm base Base of arg.
+// In erroneous situations, e.g., integer_log(base, 0) the function
+// returns 0 and does not report the error. This is the intended
+// behavior.
+inline std::size_t integer_log(std::size_t base, std::size_t arg)
+{
+  std::size_t ilog = 0;
+  while( base <= arg )
+  {
+    arg /= base; ++ilog;
+  }
+  return ilog;
+}
+
+inline std::size_t integer_pow(std::size_t base, std::size_t e)
+{
+  // Reuse the implementation of exponentiation by squaring, but
+  // there is a twist that powm actually computes base^e % m,
+  // so for m we plug-in a fully set bitset, whose modulo will be no-op.
+  return multiprecision::powm(base, e, ~std::size_t());
```

Comment:
Maybe my previous advice to use `powm` was incorrect, if you don't need to use the modulus operator. Do you think it'll compile down to a no-op? If it does, will it do the wrong thing on overflow? (I just feel like there should be some way for code reuse on this one in boost or the stdlib, maybe that's naive. . . )

Reply:
I think it's possible to reuse code there, but it will require a bit of effort.

`powm` could probably be implemented like this:

```C++
template <class I1, class I2, class I3>
typename enable_if_c<is_integral<I1>::value && is_unsigned<I2>::value && is_integral<I3>::value, I1>::type
   powm(const I1& a, I2 b, I3 c)
{
   typedef typename detail::double_integer<I1>::type double_type;

   I1 x(1), y(a);
   double_type result;

   while(b > 0)
   {
      if(b & 1)
      {
         multiply(result, x, y);
         x = integer_modulus(result, c);
      }
      multiply(result, y, y);
      y = integer_modulus(result, c);
      b >>= 1;
   }
   return integer_modulus(x, c);
}
```

(Note `return integer_modulus(x, c);` instead of plain `return x % c;`)

And if  integer_modulus would support mpl::false_() as the second argument, it could return just the first one. Which would make everything work.

Reply:
I will restore my previous code here, but let's keep in minds that with the suggestion above code reuse would be possible.